### ISSUE TITLE
Efficiency of surrogates for higher dim

### ISSUE THREAD

**Issue by agrayver** [beniz/libcmaes#106]
I took examples/surrogates/sample-code-surrogate-rsvm.cc and set dim=30, alg=acmaes, fixed seed=12345. Then, running with set_exploit(true) and set_exploit(false) one sees (figure attached) that exploitation not only does not help, but essentially breaks convergence. For low dimensions, e.g. dim=5-15, it works well however. Any idea how to make it efficient for higher dim?
![surr](https://cloud.githubusercontent.com/assets/8438035/5891394/4c791106-a493-11e4-83fe-684abc00b714.png)

**Comment by beniz** [comment]
Are you running
```
./sample_code_surrogate_rsvm -dim 30 -alg acmaes
```
? (which implicitly runs on fsphere btw). 

First try a higher initial sigma, such as:
```
./sample_code_surrogate_rsvm -dim 30 -alg acmaes -sigma0 1
```
which does the trick for me here, and beats the algorithm without surrogates and with same initial sigma.

I guess you have looked at it, but just in case, see https://github.com/beniz/libcmaes/wiki/Using-Surrogates-for-expensive-objective-functions as there's a dedicated Python script that in addition to regular useful values plots both the train and test error of the surrogate.

Also, look at additional options with
```
./sample_code_surrogate_rsvm --help
```

A potentially useful flag is `-l` that specifies the number of sample points the surrogate trains from. As dimension increases, you may expect a higher (exponential ?) requirement in sample points.

Finally, to my knowledge, the science of surrogates with CMA-ES is still pretty young. The most informed person to my knowledge is @loshchil. He has studied related techniques that are not available in the lib yet, such as #76.

**Comment by beniz** [comment]
@nikohansen there's the obvious question here of TPA + surrogates :)

**Comment by nikohansen** [comment]
What exactly does `set_exploit(true)` imply? Generally, I would consider failure of convergence on the sphere to be a bug, no matter what.

**Comment by beniz** [comment]
`set_exploit` activates the exploitation of the surrogate, that is replaces objective function calls with calls to the surrogate. It could be a bug, as a first step, I'd need to re-dig into the papers in order to compare with published results to check for discrepancy.

**Comment by beniz** [comment]
> I would consider failure of convergence on the sphere to be a bug

Understood, a good rule for the lib to follow. FTR, below is a run of 
```
./sample_code_surrogate_rsvm -dim 30 -alg acmaes -ftarget 1e-8
```
The run terminates on 'conditionCov':

![acmaes_surr_30d_fsphere](https://cloud.githubusercontent.com/assets/3530657/5891779/79c559f4-a4a6-11e4-9e7e-e0d9fa5b3c61.png)

**Comment by nikohansen** [comment]
It looks like a problem connected to step-size control though, because we see standard deviations and eigenvalues increase systematically and far above 1. This suggest that the step-size is (far) too small, which should not happen under normal circumstances.

**Comment by agrayver** [comment]
As far as I understand, the problem is in the library, isn't it? If not, is there anything a user can do about controlling step-size?

**Comment by nikohansen** [comment]
Agreed, the user can't really do anything to improve step-size control. What we see is probably a bug in the library, or a (subtle) problem from the coupling of surrogate and CMA.

**Comment by beniz** [comment]
> What we see is probably a bug in the library, or a (subtle) problem from the coupling of surrogate and CMA.

I wouldn't rule out a bug indeed. It will take me some time to re-run comparisons to the little witness runs I have from publications. For some reasons, the cluster I used to rely on is now significantly slower than my laptop, and I am trying to work around it.

**Comment by beniz** [comment]
Just an update on this issue: I do confirm since is due to several intertwined bugs. I've put a long fight and they should be fixed now, along with better default settings of a few parameters.

FYI, what I would now consider to have been the main bug, was messing the final ranking of candidates before the update.

I plan to update the original ticket #57 with new reports from runs vs published literature, along with commits.

The code has still a few rough angles, so I will mention it here once it is ready to use.

Here is the fixed run with active-CMA-ES and surrogate in 30-D on fsphere:
![surr_acmaes_fsphere_30d_fix](https://cloud.githubusercontent.com/assets/3530657/5919662/32920104-a636-11e4-9f1b-4f2f3b38b8c6.png)

**Comment by agrayver** [comment]
Great news! Looking forward to testing new version.

**Comment by beniz** [comment]
Fix as been pushed, 
```
./sample_code_surrogate_rsvm -dim 30 -alg acmaes
```
now yields the desired behavior.

Thanks for reporting the initial problem, it did indeed lead to very necessary bug fixes and improvements.

**Comment by beniz** [comment]
FYI, not all runs appear to converge equally well, and this is something I am investigating, along other improvements.

**Comment by agrayver** [comment]
Thank for information. I have not tested it yet. Please let me know when you're done, I will update and run it here.

**Comment by beniz** [comment]
Well, in fact there was no problem but an old version of the program in the path on my laptop. On the bright side, this had me check on three different machines and compilers, many runs with 100% success, so a false alarm definitely!

**Comment by agrayver** [comment]
When I do make install it seems the following files are missing in the $PREFIX:

surrogates/rankingsvm.hpp
surrogates/rsvm_surr_strategy.hpp
opti_err.h

**Comment by agrayver** [comment]
For my application, I still do not see any improvement when using surrogates. I run 40-D problem and my F-targer ~ 3000. 

Here are results with exploitation (just used RSVMSurrogateStrategy and set_exploit, all parameters are default):
![exploit](https://cloud.githubusercontent.com/assets/8438035/5965840/85171426-a7ff-11e4-9b86-37e741cd5a5d.png)

Why does it show 41-D btw?

Here are results without surrogates:
![no_exploit](https://cloud.githubusercontent.com/assets/8438035/5965855/a2e90374-a7ff-11e4-97e1-9df361ee2222.png)

It converges to the target quite fast. 
Any idea?

**Comment by beniz** [comment]
> Why does it show 41-D btw?

try using tests/cma_multiplt_surr.py to plot the results

**Comment by beniz** [comment]
> For my application, I still do not see any improvement when using surrogates. I run 40-D problem and my F-targer ~ 3000.

First, can you confirm that the previous example with fsphere and acmaes does work properly ? (just making sure the proper version of the code is running to get this out of the way).

My understanding is that you are pioneering surrogates with CMA and 'unknown' (i.e. out of benchmarks) functions, that is, in addition to potential bugs. 

From there and the graph with the right python script above, we'll be able to move forward.

**Comment by nikohansen** [comment]
> try using tests/cma_multiplt_surr.py to plot the results. This is required to plot surrogate data output.

somewhat unfortunate user interface ;-)

> For my application, I still do not see any improvement when using surrogates.

I have seen this before. Surrogates are no guaranty for an improvement, but the shown results are not (yet) conclusive. Just to cross-check: are you sure the true function values are shown in the surrogate case? 

Commenting the results without surrogate: 

- It hasn't converged yet, I guess not even close (same with surrogate). 

- It would be quite helpful to see the difference to the best function value, to see the current changes in function value (as implemented in the original plotting functions). This would reveal that there are still changes and I would expect even more significant improvements after some more adaptation has taken place. 

- We can nicely observe that a bunch of parameters is correlated, as they move up all alike at around 11000 f-evaluations. This movement corresponds to the emergence of a long axis in the eigenvalues plot. It might be insightful from the application view point to see to which parameters this corresponds. It is quite likely that we see the same in the surrogate plot, where the standard deviations of a few parameters increase in lockstep at the very end. 

- It might be useful to find out which parameter corresponds to the lowest line in the lower right plot (for this reason variable annotations are quite practical). Rescaling of this parameter accordingly should reduce the time to find a better solution (this is exactly what CMA is doing, but it takes some time to learn).

**Comment by agrayver** [comment]
Thanks for your help. I will do all the tests anew in a more systematic way, but maybe solving this installation problem first is a good idea, because it is potential cause of problems for a user:

https://github.com/beniz/libcmaes/issues/106#issuecomment-72092387

I had to copy these files manually and could make a mistake.

**Comment by beniz** [comment]
> I had to copy these files manually and could make a mistake.

This has been fixed. It is recommended you use
```
make uninstall
```
in between two versions.

**Comment by beniz** [comment]
> Just to cross-check: are you sure the true function values are shown in the surrogate case?

Certainly not, this is due to not using the dedicated script (https://github.com/beniz/libcmaes/wiki/Using-Surrogates-for-expensive-objective-functions). I will work on making a single script instead, but basically the default output function is different whether surrogate is active or not. As for now, running the correct script on the existing output should generate a proper plot.

**Comment by beniz** [comment]
> (for this reason variable annotations are quite practical)

btw, well noted, part of #91, still a lot to do.

**Comment by beniz** [comment]
> Rescaling of this parameter accordingly should reduce the time to find a better solution

@agrayver see this page about the (semi)-automated rescaling of parameters, especially if you happen to have bounded parameters (and known bounds!): https://github.com/beniz/libcmaes/wiki/Using-parameter-space-transforms-known-as-genotype-phenotype-transforms

**Comment by agrayver** [comment]
> see this page about the (semi)-automated rescaling of parameters, 
> especially if you happen to have bounded parameters (and known bounds!):

I use this, i.e. my gp is:
GenoPheno<pwqBoundStrategy, linScalingStrategy> gp(lbounds, ubounds, x0.size());

All results are reported using it. 
Still need to rerun tests from scratch. Will report asap.

**Comment by beniz** [comment]
@nikohansen as I had mentioned a while ago in https://github.com/beniz/libcmaes/issues/57#issuecomment-58200173, surrogate is learning from genotype. This had not triggered a discussion at the time.

**Comment by beniz** [comment]
> Still need to rerun tests from scratch. Will report asap.

Some of my new tests take a long time to run, but I should let you know that in higher dims, there are results from the literature I cannot yet fully reproduce, which indicates something is off and/or some parameters need more tweaking.

**Comment by nikohansen** [comment]
Re annotations: one relatively simple way to get all the goodies from the existing plotting routines would be to convert the output file to a format the existing plotting routines can read (one file for each subfigure). This should be relatively straight forward.

**Comment by beniz** [comment]
@agrayver OK, so last fix does the trick, results now on par with literature. Sorry for the inconvenience, this was a nasty and well hidden one. This should now run both faster and perform better.

**Comment by beniz** [comment]
@nikohansen good idea, let's work this as #110

**Comment by agrayver** [comment]
I cannot compile freshly installed RSVM-related code. From surrcmaes.h these includes do not exist:

    #include "surrogates/rankingsvm.hpp"
    #include "surrogates/rsvm_surr_strategy.hpp"

changing to 

    #include "rankingsvm.hpp"
    #include "rsvm_surr_strategy.hpp"

works however.

**Comment by beniz** [comment]
@agrayver OK, thanks for the heads up, I know where this comes from, assuming you are using `make install`.

**Comment by agrayver** [comment]
Could you please confirm that what I get below is the correct behaviour?

     ./sample_code_surrogate_rsvm -dim 40 -alg acmaes -fplot output.dat -no_exploit
![fsphere_40d_noexp](https://cloud.githubusercontent.com/assets/8438035/5987802/d7f0b296-a940-11e4-859a-d5ad61a5fb0d.png)

     ./sample_code_surrogate_rsvm -dim 40 -alg acmaes -fplot output.dat
![fshpere_40d](https://cloud.githubusercontent.com/assets/8438035/5987805/e55d835a-a940-11e4-8ed4-94009887c722.png)

**Comment by agrayver** [comment]
For 10D Rosenbrock surrogates do not seem to help:

    -dim 10 -alg acmaes -fplot output.dat -fname rosenbrock -no_exploit
![rosenbrock_10d_noexp](https://cloud.githubusercontent.com/assets/8438035/5987838/48f9ac76-a942-11e4-94ee-a8715de7f8ad.png)

    -dim 10 -alg acmaes -fplot output.dat -fname rosenbrock
![rosenbrock_10d](https://cloud.githubusercontent.com/assets/8438035/5987839/51e96a10-a942-11e4-8efa-6fee22570068.png)

**Comment by beniz** [comment]
@agrayver yes, it does. To 1e-8, I get 5565 f-evals without surrogates, and 2815 with surrogate exploitation.

**Comment by beniz** [comment]
@agrayver below are the results for Rosenbrock (that I still need to add to #57), average fevals over 10 runs with std deviation and number of successful runs:

```
running surrogates on rosenbrock
D       fevals_avg
2       573.667 +/- 357.795 (9)
4       952.667 +/- 360.211 (9)
5       1072.38 +/- 594.202 (8)
8       1384.1 +/- 111.025 (10)
10      1895.67 +/- 221.716 (9)
16      5556.33 +/- 1014.05 (9)
20      8150.4 +/- 681.513 (10)
32      27046 +/- 1550.77 (9)
40      41545 +/- 2852.78 (9)
```

This uses the tests/surr_test exe and default parameters are not the same as in the sample code.

**Comment by agrayver** [comment]
@beniz would you mind please elaborate which parameters you think need to be tweaked for the sample code to be more efficient?

**Comment by beniz** [comment]
@agrayver try setting same x0 on runs you want to compare, and for Rosenbrock, you can use `-l` with value `~40 * sqrt(N)`, i.e. ~ 210 in 10-D.

**Comment by beniz** [comment]
Typically, compare (no exploitation) 
```
./sample_code_surrogate_rsvm -ftarget 1e-8 -dim 10 -alg acmaes -fname rosenbrock -fplot rosen_noex.dat -l 210 -x0 2 -no_exploit
```
to (exploitation)
```
./sample_code_surrogate_rsvm -ftarget 1e-8 -dim 10 -alg acmaes -fname rosenbrock -fplot rosen_ex.dat -l 210 -x0 2
```

**Comment by agrayver** [comment]
Thank you for the help. Using -l ~70 * sqrt(N) seems to work. I see significant reduction (1.5-3x) of function calls for different functions and dimension (still have not tested my real app). At the moment I am testing 100D Rosenbrock and it takes very long time. It's been running for more than 5 hours on my quite modern laptop and it is still far from finishing. I'm hence wondering:

1. What is the complexity of the RSVM with respect to N?
2. What exactly takes so much time, is there potential for optimization?

**Comment by beniz** [comment]
Try lowering the number of iterations of the RSVM algo 'rsvm_iter' to something in between 150000 and 1M which is the current very conservative default.

**Comment by beniz** [comment]
> What is the complexity of the RSVM with respect to N?

Roughly speaking, O(niter,l-1) for the algorithm, with l~70*sqrt(N), and O(N*l^2) for the kernel computation.

For thorough details, see https://www.lri.fr/~ilya/phd.html page 81 (section 4.1.1)

**Comment by agrayver** [comment]
Thank you, I got a reasonable performance with testing function. Now, switching to my actual problem I still see that surrogates deteriorate convergence. Here is results with no exploitation, 40D:

![mt_40d_noexp](https://cloud.githubusercontent.com/assets/8438035/6011070/5e285b22-ab38-11e4-91f3-55b7407ed5da.png)

Here RSVM was used (with -l 500 and rsvm_iter 200000):

![mt_40d_exp](https://cloud.githubusercontent.com/assets/8438035/6011078/7925f66e-ab38-11e4-8a6d-be5871ce0b23.png)

I used the same starting guess and seed for both.
Any idea?

**Comment by beniz** [comment]
At first glance I d suggest you try increasing rsvm_iter and see if at least it does improve convergence.

On February 3, 2015 12:06:34 AM GMT+01:00, Alexander Grayver <notifications@github.com> wrote:
>Thank you, I got a reasonable performance with testing function. Now,
>switching to my actual problem I still see that surrogates deteriorate
>convergence. Here is results with no exploitation, 40D:
>
>![mt_40d_noexp](https://cloud.githubusercontent.com/assets/8438035/6011070/5e285b22-ab38-11e4-91f3-55b7407ed5da.png)
>
>Here RSVM was used (with -l 500 and rsvm_iter 200000):
>
>![mt_40d_exp](https://cloud.githubusercontent.com/assets/8438035/6011078/7925f66e-ab38-11e4-8a6d-be5871ce0b23.png)
>
>I used the same starting guess and seed for both.
>Any idea? 
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/beniz/libcmaes/issues/106#issuecomment-72558323

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by nikohansen** [comment]
It looks like the initial step-size is at least a factor 1000 (in the previous plot a factor 1e5) too small and the mechanism to prevent one eigenvalue to massively increase doesn't work out here. The second point should be checked in the lib.

**Comment by beniz** [comment]
> The second point should be checked in the lib.

This is not utterly clear to me, could you give me more details, thanks!

**Comment by nikohansen** [comment]
Sorry, yes, this is the factor `hsig` which "modulates" the learning rate for the update of the evolution path `pc` for the covariance matrix update. Looking at the data carefully reveals however that the step-size `sigma` is increasing very quickly only in the beginning, where `hsig` seems to be effective. That means the reason is not likely to be a mistaken computation of `hsig`, but a too slow increment of `sigma`. It doesn't look like ideal strategy behavior.

**Comment by beniz** [comment]
Closing for no activity after 15 days. Can be reopened as needed.
